src/libsyntax/ext/expand.rs

```diff
@@ -531,11 +532,35 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         let path = &mac.node.path;
 
         let ident = ident.unwrap_or_else(|| keywords::Invalid.ident());
-        let validate_and_set_expn_info = |def_site_span,
+        let validate_and_set_expn_info = |self_: &mut Self, // arg instead of capture
+                                          def_site_span,
                                           allow_internal_unstable,
-                                          allow_internal_unsafe| {
+                                          allow_internal_unsafe,
+                                          // can't infer this type
+                                          unstable_feature: Option<(Symbol, u32)>| {
+
+            // feature-gate the macro invocation
+            if let Some((feature, issue)) = unstable_feature {
+                let crate_span = self_.cx.current_expansion.crate_span.unwrap();
+                // don't stability-check macros in the same crate
+                if !crate_span.contains(def_site_span)
+                    && !span.allows_unstable() && self_.cx.ecfg.features.map_or(true, |feats| {
+                    // macro features will count as lib features
+                    !feats.declared_lib_features.iter().any(|&(feat, _)| feat == feature)
+                }) {
+                    let explain = format!("macro {}! is unstable", path);
```

Comment:
If you are checking stability, then you can check deprecation as well - these two properties are normally checked together.
IIRC, there was even a separate issue about deprecation attributes being ignored on macros.

Reply:
That's doable, except I don't know how to get the crate's `#[warn/allow/deny(deprecation)]` flag in this context.

Reply:
I guess we have to separately find and store that value as we walk the AST while expanding macros. It's doable if we want this feature.

Reply:
Is this not already tracked somewhere in the AST, like for the dead code lints? Or is macro expansion too early for that?

Reply:
Deprecation tracking is done in phase 3 at the same time as stability tracking. I'm having to more-or-less naively reimplement the latter here (unless we want to do what @petrochenkov suggested and refactor stability checking entirely into phase 2) so deprecation tracking is going to have to be reimplemented as well.

Reply:
Ok, if there are complications then let's leave this for later.